### DISCUSSION TITLE
Setup UI / UX

### DISCUSSION THREAD

**Discussion by jkitman** [fedimint/fedimint#2212] [General]
A Fedimint consists of a group of individuals called “guardians” who need to perform a setup process so they all have the same configuration.  We can discuss ideas on how best to design the user experience, but here is my first draft at it:

**First the guardians need to coordinate over a secure communication channel such as Signal, Matrix, or face-to-face**
    • The guardians agree on who will be allowed into the federation
    • Each guardian deploys their own Fedimint server (could be through a VPS such as Digital Ocean or a self-hosted node package such as Umbrel)
    • The deployment process should let the guardian know what their UI URL is (for instance https://209.85.255.255 or https://gaurdian-alpha.xyz)

**Next, all guardians visit the URL which brings them to the Setup UI where they must all connect to each other:**
    • Guardians must enter a password that would be required every time they log into the Admin UI (can choose whether password is also stored on the server)
    • Guardians must enter a “server name” that would make it easier for other guardians to identify them
    • Guardians can view their API URL and must decide on a “setup leader”
    • The leader shares their API URL with all other “setup follower” guardians
    • Followers enter the leader’s API URL into their UI
    • During this time, all guardians can see all other guardians (identified by names and API URLs) connecting to the leader

**The leader configures the federation for all guardians**
    • As guardians connect to the leader, the option becomes available for the leader to modify configuration values
    • Modules can be added or removed
    • Every module has its own set of configuration values which will be the key → value format
    • Modules can start populated with default recommended values (generally these won’t be changed)
    • Once all modules are configured AND all guardians have connected the leader “finalizes” the config
    • After finalizing the config, the “key generation” step runs which creates private keys for each guardian (could take around a minute)
    • While waiting for key generation, followers should be able to view the finalized config received from the leader

**Guardians enter the config verification stage**
    • Using a secure communication channel, all guardians must share a verification object with each other
    • The verification object could be a text string, emojis, image pattern (open to ideas on the best option here)
    • Guardians must pass a challenge test in the UI to prove they received the correct verification objects from each other
    • The challenge test could be a multiple-choice selection or copy-and-paste text (open to ideas on the best option here)
    • Guardians can see updates of other guardians passing the verification stage
    • Once all guardians pass, Fedimint consensus starts and we should load the Admin UI

**Comment by garyray-k** [comment]
1. Will a more robust authentication method be supported in the future? i.e. ssh keys or something similar, 2FA?
2. Will this all be GUI based or will there be a CLI administration option? I'd much rather administer my server via CLI as a more technical person. However, I realize this level of comfort may not be expected and the point-and-click method is a must for wide adoption.
>After finalizing the config, the “key generation” step runs which creates private keys for each guardian (could take around a minute) 
3. Are these keys stored on the server? Any option for a hardware wallet or other key management?

I assume this is MVP so apologies if these questions were considered but not in the first draft.

**Reply by elsirion** [reply]
> will there be a CLI administration option

While it isn't a focus of this document, a major improvement of the setup process we are working on is making it API based and the Web UI just one possible way of interacting with it. So it will be possible to have a CLI or a dedicated app for these operations.

For now you can also still use the `distributedgen` CLI tool to manually run the config generation from your CLI and if there already is config present the config gen server won't even start.

> Will a more robust authentication method be supported in the future? i.e. ssh keys or something similar, 2FA?

I'd be happy about suggestions, I could imagine having other API authentication mechanisms in the future. I haven't really worked with "web techologies" for some time, so there might be suitable standards we weren't aware of.

> Are these keys stored on the server? Any option for a hardware wallet or other key management?

Not for now and that was a very conscious choice (learning from other projects). HSMs are a huge operational hassle and if implemented incorrectly don't even have any security benefits. For an HSM to be useful it would need to be able to independently verify the system's state, i.e. essentially run the whole consensus, which is somewhat impractical.

**Reply by jkitman** [reply]
Good questions.

1. The security model of Fedimint is "nFA" where `n` is 2/3+1 of the nodes :smiley:  But yes we could also eventually support a hardware or Auth app based key in the future.  Another secure option is to run the UI only as a local server which could contain a secret that is hashed with the password.
2. Most of the API commands will be supported via a remote CLI, although some things like upgrades may need to be run on the server itself.  The deployment method would determine how best to login, but probably something like SSH.
3. Keys need to be stored on the server because the consensus logic is quite complex and I can't imagine how that would be ported to a hardware wallet.  However, we do encrypt the keys at-rest.  Maybe there's a way we could support a cold wallet for manually storing some of the on-chain funds, similar to how exchanges do this.
